coprocessor/fhevm-engine/transaction-sender/src/nonce_managed_provider.rs

```diff
@@ -59,6 +63,84 @@ impl<P: alloy::providers::Provider<Ethereum> + Clone + 'static> NonceManagedProv
         res
     }
 
+    pub async fn send_transaction_sync(
+        &self,
+        tx: impl Into<TransactionRequest>,
+        timeout: Duration,
+    ) -> TransportResult<TransactionReceipt> {
+        let mut tx = tx.into();
+        if let Some(signer_address) = self.signer_address {
+            let nonce_manager = self.nonce_manager.lock().await;
+            let nonce = nonce_manager
+                .get_next_nonce(&self.provider, signer_address)
+                .await?;
+            tx.nonce = Some(nonce);
+        }
+
+        let res = tokio::time::timeout(timeout, self.provider.send_transaction_sync(tx))
+            .await
+            .map_err(|_| TransportErrorKind::custom_str("eth_sendRawTransactionSync timeout"))
+            .flatten();
+
+        if res.is_err() {
+            // Reset the nonce manager if the transaction sending failed.
```

Comment:
This is where i see the biggest flaw in the  implementation. 
Because in case of error, you are reseting the nonce automatically. This is reducing the capacity of parallelism and introduce a lot of nonce too high errors. 
Could be resolved in a next PR, by using a different nonce management strategy.